### PR TITLE
Updates for Clevis changes to API

### PR DESCRIPTION
**Pull request by mulkieran**
Related https://github.com/stratis-storage/project/issues/275

### CODE REVIEW

**Comment by mulkieran** [comment]
Should fail lots of stratis-cli tests.

**Comment by mulkieran** [comment]
@drckeefe How do you want the pool create command to appear now that it can be bound w/ clevis on encryption?

Help text for create:

```
usage: stratis pool create [-h] [--redundancy {none}] [--key-desc KEY_DESC]
                           pool_name blockdevs [blockdevs ...]

positional arguments:
  pool_name            Name of new pool
  blockdevs            Create the pool using these block devs

optional arguments:
  -h, --help           show this help message and exit
  --redundancy {none}  Redundancy level for this pool
  --key-desc KEY_DESC  Key description of key in kernel keyring to use for
                       encryption
```

Help text for bind:

```
usage: stratis pool bind [-h] {nbde,tang,tpm2} ...

optional arguments:
  -h, --help        show this help message and exit

subcommands:
  {nbde,tang,tpm2}
    nbde (tang)     Bind using NBDE via a tang server
    tpm2            Bind using TPM2
```

**Comment by mulkieran** [comment]
The point is that the tpm2 and tang subcommands are rather different:

```
usage: stratis pool bind nbde [-h] (--trust-url | --thumbprint THUMBPRINT)
                              pool_name key_description url

positional arguments:
  pool_name             Pool name
  key_description       Description of key in kernel keyring used by this pool
  url                   URL of tang server

optional arguments:
  -h, --help            show this help message and exit
  --trust-url           Omit verification of tang server credentials
  --thumbprint THUMBPRINT
                        Thumbprint of tang server at specified URL
```
and 
```
usage: stratis pool bind tpm2 [-h] pool_name key_description

positional arguments:
  pool_name        Pool name
  key_description  Description of key in kernel keyring used by this pool

optional arguments:
  -h, --help       show this help message and exit
```

**Comment by drckeefe** [comment]
Here is a possible option for Clevis in pool create

stratis pool create <pool name> --key-desc <key> --bind [tpm2 |[nbde|tang (<thumbprint>|--trust-url) <url>] <device(s)>

examples:
Bind to TPM2:
# stratis pool create p1 --key-desc key1 --bind tpm2  /dev/md0
Bind to Tang with thumbprint:
# stratis pool create p1 --key-desc key1 --bind tang 2io2ml22io3osdofio23 http://tang.corp /dev/md0
Bind to Tang without thumbprint: 
# stratis pool create p1 --key-desc key1 --bind tang --trust-url http://tang.corp /dev/md0

**Comment by mulkieran** [comment]
rebased.

**Comment by drckeefe** [comment]
I'm not opposed to --clevis, though we have a bind option and clevis would be new to the CLI.  I also believe that bind is a generic term that more users would be aware of.

**Comment by mulkieran** [comment]
> I'm not opposed to --clevis, though we have a bind option and clevis would be new to the CLI. I also believe that bind is a generic term that more users would be aware of.

```clevis``` not entirely new. Consider "unlock" command.

**Comment by drckeefe** [comment]
Ah, true!

**Comment by mulkieran** [comment]
Current version of help text for ```pool create```

```
usage: stratis pool create [-h] [--redundancy {none}] [--key-desc KEY_DESC]
                           [--clevis {nbde,tang,tpm2}] [--tang-url URL]
                           [--trust-url | --thumbprint THUMBPRINT]
                           pool_name blockdevs [blockdevs ...]

positional arguments:
  pool_name             Name of new pool
  blockdevs             Create the pool using these block devs

optional arguments:
  -h, --help            show this help message and exit
  --redundancy {none}   Redundancy level for this pool
  --key-desc KEY_DESC   Key description of key in kernel keyring to use for
                        encryption
  --clevis {nbde,tang,tpm2}
                        Specification for binding with Clevis.
  --tang-url URL        URL of Clevis tang server (ignored if
                        --clevis=[tang|nbde] not set)
  --trust-url           Omit verification of tang server credentials (ignored
                        if --clevis=[tang|nbde] not set)
  --thumbprint THUMBPRINT
                        Thumbprint of tang server at specified URL (ignored if
                        --clevis=[tang|nbde] not set)
```

**Comment by mulkieran** [comment]
@drckeefe Please take a look at the implementation and try running it. You should be able to run the extended ```pool create``` command to find all the new options. The implementation choice I've made involves one more option than you asked for ```--thumbprint``` this makes sense for the parser.

A consequence of the choice is that the parser library can not handle the complexity of the restrictions that we need to handle, and so some invalid command-lines are parsed by the parser and then an exception indicating they are invalid is raised later.

An example of this is ```stratis pool create --clevis=tang polname /dev/vde```. Note that no URL is specified, which is wrong. This is treated as an invalid, but not an ungrammatical command-line, the error code is 1.

However, ```stratis pool create --clevis=tang --trust-url -thumbprint=jkajfkj``` is considered ungrammatical, because the parser is able to identify that ```--trust-url``` and ```--thumbprint=jljl``` don't make sense together, the error code is 2. The same combination is also considered ungrammatical in the ```bind``` command.

Try them and see. The implementation does not have an error if the user has set some additional irrelevant options like:
```--clevis=tmp2 --trust-url```, the ```--trust-url``` is just ignored. If you want an error on that, it's fairly easy to arrange.

**Comment by mulkieran** [comment]
I think that the D-Bus API shared method is just not unmarshalling the clevis info, which explains why all the tests are passing.

File: src/stratis_cli/_actions/_utils.py [hunk 1]
```diff
@@ -92,3 +104,47 @@ def fetch_property(props, name):
...
+            clevis_config = {CLEVIS_KEY_URL: namespace.tang_url}
+            if not namespace.trust_url and namespace.thumbprint is None:
+                raise StratisCliMissingClevisThumbprintError()
+
+            if (
+                # pylint: disable=bad-continuation
+                namespace.trust_url
+                and not namespace.thumbprint is None
+            ):  # pragma: no cover
+                raise AssertionError("--trust-url and --thumbprint both set")
```
Comment (jbaublitz) [inline src/stratis_cli/_actions/_utils.py:133]: Is this necessary given that these arguments are specified as mutually exclusive in the parser?

**Review by jbaublitz — COMMENTED** [review]
Just one question about the parser validation.

File: src/stratis_cli/_actions/_utils.py (hunk 1 above)
Comment (mulkieran) [reply src/stratis_cli/_actions/_utils.py]: It's trying to hold the spot that an ```unreachable``` would in Rust. I've tidied it up in the next commit.

**Comment by mulkieran** [comment]
New help text for unbind:

```
usage: stratis pool unbind [-h] pool_name {keyring,clevis}

positional arguments:
  pool_name         Pool name
  {keyring,clevis}  Encryption method to unbind

optional arguments:
  -h, --help        show this help message and exit
```

```clevis``` is the default, so does not need to be specified.

**Comment by mulkieran** [comment]
New help text for bind:

```
usage: stratis pool bind [-h] {nbde,tang,tpm2,keyring} ...

optional arguments:
  -h, --help            show this help message and exit

subcommands:
  {nbde,tang,tpm2,keyring}
    nbde (tang)         Bind using NBDE via a tang server
    tpm2                Bind using TPM2
    keyring             Bind using the kernel keyring
```
Now has one more subcommand, ```keyring``` which has standard arguments.

```
usage: stratis pool bind keyring [-h]
                                 (--keyfile-path KEYFILE_PATH | --capture-key)
                                 pool_name keydesc

positional arguments:
  pool_name             Pool name
  keydesc               key description

optional arguments:
  -h, --help            show this help message and exit
  --keyfile-path KEYFILE_PATH
                        Path to the key file containing a key to set in the
                        keyring
  --capture-key         Read key from stdin with no terminal echo or userspace
                        buffer storage
```

**Comment by mulkieran** [comment]
default test errors are all acceptable.

**Comment by drckeefe** [comment]
> New help text for unbind:
> 
> ```
> usage: stratis pool unbind [-h] pool_name {keyring,clevis}
> 
> positional arguments:
>   pool_name         Pool name
>   {keyring,clevis}  Encryption method to unbind
> 
> optional arguments:
>   -h, --help        show this help message and exit
> ```
> 
> Alternate help text for unbind, making the method an option, instead.
> 
> ```
> usage: stratis pool unbind [-h] [--method {keyring,clevis}] pool_name
> 
> positional arguments:
>   pool_name             Pool name
> 
> optional arguments:
>   -h, --help            show this help message and exit
>   --method {keyring,clevis}
>                         Encryption method to unbind
> ```
> 
> The advantage of the second is that there is a default, if the user wants to unbind from clevis, which we think is the usual case, they can just call `stratis unbind <poolname>` and be done. We need to decide which is best. I've implemented the second.

**Comment by drckeefe** [comment]
> > New help text for unbind:
> > ```
> > usage: stratis pool unbind [-h] pool_name {keyring,clevis}
> > 
> > positional arguments:
> >   pool_name         Pool name
> >   {keyring,clevis}  Encryption method to unbind
> > 
> > optional arguments:
> >   -h, --help        show this help message and exit
> > ```
> > 
> > 
> > Alternate help text for unbind, making the method an option, instead.
> > ```
> > usage: stratis pool unbind [-h] [--method {keyring,clevis}] pool_name
> > 
> > positional arguments:
> >   pool_name             Pool name
> > 
> > optional arguments:
> >   -h, --help            show this help message and exit
> >   --method {keyring,clevis}
> >                         Encryption method to unbind
> > ```
> > 
> > 
> > The advantage of the second is that there is a default, if the user wants to unbind from clevis, which we think is the usual case, they can just call `stratis unbind <poolname>` and be done. We need to decide which is best. I've implemented the second.

I would prefer no defaults on unbind.  I think we should ask the user to be specific about their choice here.

**Comment by drckeefe** [comment]
> @drckeefe Please take a look at the implementation and try running it. You should be able to run the extended `pool create` command to find all the new options. The implementation choice I've made involves one more option than you asked for `--thumbprint` this makes sense for the parser.
> 
> A consequence of the choice is that the parser library can not handle the complexity of the restrictions that we need to handle, and so some invalid command-lines are parsed by the parser and then an exception indicating they are invalid is raised later.
> 
> An example of this is `stratis pool create --clevis=tang polname /dev/vde`. Note that no URL is specified, which is wrong. This is treated as an invalid, but not an ungrammatical command-line, the error code is 1.
> 
> However, `stratis pool create --clevis=tang --trust-url -thumbprint=jkajfkj` is considered ungrammatical, because the parser is able to identify that `--trust-url` and `--thumbprint=jljl` don't make sense together, the error code is 2. The same combination is also considered ungrammatical in the `bind` command.
> 
> Try them and see. The implementation does not have an error if the user has set some additional irrelevant options like:
> `--clevis=tmp2 --trust-url`, the `--trust-url` is just ignored. If you want an error on that, it's fairly easy to arrange.

Reviewing and sent some comments to you @mulkieran

**Comment by mulkieran** [comment]
> > > New help text for unbind:
> > > ```
> > > usage: stratis pool unbind [-h] pool_name {keyring,clevis}
> > > 
> > > positional arguments:
> > >   pool_name         Pool name
> > >   {keyring,clevis}  Encryption method to unbind
> > > 
> > > optional arguments:
> > >   -h, --help        show this help message and exit
> > > ```
> > > 
> > > 
> > > Alternate help text for unbind, making the method an option, instead.
> > > ```
> > > usage: stratis pool unbind [-h] [--method {keyring,clevis}] pool_name
> > > 
> > > positional arguments:
> > >   pool_name             Pool name
> > > 
> > > optional arguments:
> > >   -h, --help            show this help message and exit
> > >   --method {keyring,clevis}
> > >                         Encryption method to unbind
> > > ```
> > > 
> > > 
> > > The advantage of the second is that there is a default, if the user wants to unbind from clevis, which we think is the usual case, they can just call `stratis unbind <poolname>` and be done. We need to decide which is best. I've implemented the second.
> 
> I would prefer no defaults on unbind. I think we should ask the user to be specific about their choice here.

That is easy to arrange. @bgurney-rh @jbaublitz Is making the unbind method a positional argument fine with you as well?

**Comment by bgurney-rh** [comment]
> > > > New help text for unbind:
> > > > ```
> > > > usage: stratis pool unbind [-h] pool_name {keyring,clevis}
> > > > 
> > > > positional arguments:
> > > >   pool_name         Pool name
> > > >   {keyring,clevis}  Encryption method to unbind
> > > > 
> > > > optional arguments:
> > > >   -h, --help        show this help message and exit
> > > > ```
> > > > 
> > > > 
> > > > Alternate help text for unbind, making the method an option, instead.
> > > > ```
> > > > usage: stratis pool unbind [-h] [--method {keyring,clevis}] pool_name
> > > > 
> > > > positional arguments:
> > > >   pool_name             Pool name
> > > > 
> > > > optional arguments:
> > > >   -h, --help            show this help message and exit
> > > >   --method {keyring,clevis}
> > > >                         Encryption method to unbind
> > > > ```
> > > > 
> > > > 
> > > > The advantage of the second is that there is a default, if the user wants to unbind from clevis, which we think is the usual case, they can just call `stratis unbind <poolname>` and be done. We need to decide which is best. I've implemented the second.
> > 
> > 
> > I would prefer no defaults on unbind. I think we should ask the user to be specific about their choice here.
> 
> That is easy to arrange. @bgurney-rh @jbaublitz Is making the unbind method a positional argument fine with you as well?

Yes, that sounds good to me.

**Comment by mulkieran** [comment]
Ok. "unbind" method is now a positional argument.

**Comment by mulkieran** [comment]
Test failures still as expected.

**Comment by mulkieran** [comment]
Ok. I've added the man pages and put in final review. The only really broad question is whether to report an error when an option that is set is simply ignored because it is irrelevant.

**Review by jbaublitz — APPROVED** [review]

**Comment by drckeefe** [comment]
I've performed functional testing for this PR.  I only found one issue around adding block devices to a pool, if the pool was created with clevis.  This issue should be resolved already or will be soon.

**Review by drckeefe — APPROVED** [review]

**Pull request closed without merging**